### PR TITLE
Add osu!taiko mod `Simplified Rhythm`

### PR DESCRIPTION
**Pull request by Hiviexd**
This mod converts 1/8 and 1/6 patterns to 1/4, and 1/3 patterns to 1/2 (all based on user preference via the mod's settings)

https://github.com/user-attachments/assets/f983291c-effb-4985-b0f7-787e2a6e355c

---

### Why should this mod (and similar mods) exist

This stands out as the first mod that alters objects via deleting them/changing their position. osu!taiko is visually a very simple game mode which makes it challenging to express creativity, which means that rhythm, coloring, and SVs are the most important variables at our disposal for creating mods for taiko, unlike osu! where you have endless possibilities with circle/slider positions, sizes, their approach circles, and so on.

In osu!taiko, using 1/6 (and 1/8 to a lesser extent), is a *very* common improvisation method in higher difficulty beatmaps, where you'd often find isolated 1/6 patterns, or have them *inside* of 1/4 patterns/streams (see the 2nd beatmap in the video). This kind of improv poses a higher barrier of entry skill-wise, so this mod aims to make these beatmaps more accessible to lower-skilled players, hence why it's a difficulty reduction mod.

### CODE REVIEW

**Comment by minisbett** [comment]
You should ensure this does not produce complications with difficulty & pp calculation.

In the standard ruleset for example, any added or removed slider would mess with calculation, not sure how it is for taiko.

**Comment by Hiviexd** [comment]
I've set up a [test beatmap](https://nats.are-la.me/4icRGj6.osz) with 2 difficulties and compared them with `osu-tools`.

Theoretically, `1/6 diff` + SR mod `1/6 diff manually converted` on NoMod should be the same exact beatmap. Here's their respective `osu-tools` difficulty info in [this gist](https://gist.github.com/Hiviexd/c933efd5d99e873e786f281d7d0d4dff).

You'll notice identical SR and *very* negligible difference in PP (`325.76` vs `325.74`), which I'm personally suspecting a small margin of error because of potential 1ms unsnaps (rounding errors).

Additionally when making the mod, there was an expected and sane SR/PP drop on every beatmap I tested, I feel pretty confident that difficulty calculation shouldn't be affected negatively by this.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                        if (!inPattern)
+                        {
+                            patternStartIndex = i;
+                        }
+
+                        inPattern = true;
+                    }
+
+                    // check if end of pattern or if we're on the last note
+                    if ((inPattern && snapValue != snapConversion.Key) || i == hits.Count)
```
Comment (Lawtrohux) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs:77]: `i == hits.Count`  cannot equal true because the loop runs while `i` is less than `hits.Count`

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+            // Snap conversions for rhythms
+            var snapConversions = new Dictionary<int, int>
+            {
+                { 8, 4 }, // 1/8 snap to 1/4 snap
+                { 6, 4 }, // 1/6 snap to 1/4 snap
+                { 3, 2 }, // 1/3 snap to 1/2 snap
+            };
+
+            int patternStartIndex = 0;
```
Comment (Lawtrohux) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs:49]: Isolate the state here by moving it into `foreach (var snapConversion in snapConversions)`. Just to make sure the variable is reset at the beginning of each iteration.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                                {
+                                    toRemove.Add(hits[j]);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // Remove queued notes
+                taikoBeatmap.HitObjects = taikoBeatmap.HitObjects.Except(toRemove).ToList();
```
Comment (Lawtrohux) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs:112]:
```suggestion
                taikoBeatmap.HitObjects.RemoveAll(obj => toRemove.Contains(obj));

```

`Except` creates a new collection, this just modifies the one in place.

**Review by Lawtrohux — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Taiko.Beatmaps;
+using osu.Game.Rulesets.Taiko.Objects;
+
+namespace osu.Game.Rulesets.Taiko.Mods
+{
+    public class TaikoModSimplified : Mod, IApplicableToBeatmap
+    {
+        public override string Name => "Simplified";
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:20]: Name feels uninspired. Alternative suggestions would be something with "quarters". Or maybe "Four On The Floor" or something (to annotate the worst you have to deal with is 1/4s).

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        public override ModType Type => ModType.DifficultyReduction;
+
+        [SettingSource("1/3 to 1/2 conversion", "Converts 1/3 patterns to 1/2 rhythm.")]
+        public Bindable<bool> OneThirdConversion { get; } = new BindableBool(false);
+
+        [SettingSource("1/6 to 1/4 conversion", "Converts 1/6 patterns to 1/4 rhythm.")]
+        public Bindable<bool> OneSixthConversion { get; } = new BindableBool(true);
+
+        [SettingSource("1/8 to 1/4 conversion", "Converts 1/8 patterns to 1/4 rhythm.")]
+        public Bindable<bool> OneEighthConversion { get; } = new BindableBool(false);
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:33]: Is this degree of control really needed?

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                }
+
+                // Remove queued notes
+                taikoBeatmap.HitObjects.RemoveAll(obj => toRemove.Contains(obj));
+            }
+        }
+
+        private int getSnapBetweenNotes(ControlPointInfo controlPointInfo, Hit currentNote, Hit nextNote)
+        {
+            double gapMs = Math.Max(currentNote.StartTime, nextNote.StartTime) - Math.Min(currentNote.StartTime, nextNote.StartTime);
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:119]: Why does this essentially take the absolute difference here, even knowing which note is "current" and which is "next"? Which is to say, why is this not just `nextNote.StartTime - currentNote.StartTime`?

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            var snapConversions = new Dictionary<int, int>
+            {
+                { 8, 4 }, // 1/8 snap to 1/4 snap
+                { 6, 4 }, // 1/6 snap to 1/4 snap
+                { 3, 2 }, // 1/3 snap to 1/2 snap
+            };
+
+            bool inPattern = false;
+
+            List<Hit> hits = taikoBeatmap.HitObjects.Where(obj => obj is Hit).Cast<Hit>().ToList();
```
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:51]: I have a feeling that doing this is incorrect, because doing something like placing a drum roll on a 1/1 beat after a 1/3rds rhythm is going to result the last note of the 1/3rds remaining snapped to 1/3. In other words, discounting non-hits this early essentially deletes non-hits from the map, skewing time intervals between objects and potentially ending patterns prematurely.

**Review by bdach — CHANGES_REQUESTED** [review]
Mod needs unit tests for correctness of applied processing

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs (hunk 6 above)
Comment (Hiviexd) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:119]: original intention was to make it agnostic of order for safety but I guess I'm accounting for an impossible case here lol

will change to what you have

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs (hunk 4 above)
Comment (Hiviexd) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:20]: thoughts on "Quarterize"?

File: osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs (hunk 7 above)
Comment (Hiviexd) [reply osu.Game.Rulesets.Taiko/Mods/TaikoModSimplified.cs:51]: I think you might be misunderstanding this.

the way the mod works is by reading patterns in groups of 3, then adjusting every 2nd note to form a 1/4 rhythm and removing the 3rd note entirely, with the mod being blind to drumrolls/swells.

the latter was a deliberate decision because having them account for conversion would result in unintended consequences, namely them getting deleted or missnapped.

here's the only 3 outcomes that can happen involving them right now:

### Case 1

| before | after |
| :-: | :-: |
| image |image |

the pattern will be untouched which is fine because snaps are not much relevant with drumrolls/swells compared to notes playability-wise.

---

### Case 2

| before | after |
| :-: | :-: |
| image | image |

the 2nd don will be moved closer to the drumroll/swell, which is arguably fine for the aforementioned reason.

---

### Case 3

| before | after |
| :-: | :-: |
| image | image |

the 2nd note will be adjusted accordingly, and the 3rd note would be removed, as expected. playability around the drumroll/swell is unaffected

**Comment by Hiviexd** [comment]
So, adding the tests has revealed an unexpected and niche issue.

The first run of every test passes, however subsequent runs of the tests that edit the objects' offsets break, mainly because the beatmap is not properly reloaded.

Important notice that this issue is only present in tests, and does not happen in regular play due to beatmaps being fully loaded/reloaded.

I've briefly talked to frenzi about this and the main gist is states are being polluted, as usually the beatmap is cloned in `GetPlayableBeatmap` but it seems the references of the hitobjects stay the same.

Apparently, the osu! Random mod manages to successfully modify an object's state so I'll look into that tomorrow, alternatively I'll look into potentially creating new hitobjects rather than editing them directly.

**Comment by DarkStoorM** [comment]
I decided to check this mod on a map with very frequent beat length changes, like [this one](https://osu.ppy.sh/beatmapsets/638476#taiko/1354418).

I found a weird section, where having 1/3 and 1/6 conversions are not working as expected (I think?)
Example section: (02:20:997). There is a 1/6 (dddk kd). In the (kd) part, I don't know what happened there, but it looks like the don got snapped to 1/12 

image

It's this particular pattern that's causing this, a 1/6 hit + double (like you'd insert a 1/6 quad and removed the second hit) and it only happens while having both 1/3 and 1/6 conversions on.

image

I know it will affect a very, **very tiny** portion of beatmaps, but I'm just wondering if that's a mistake or some combinations will just not work/are not supposed to be used together.

**Comment by Hiviexd** [comment]
Issue with tests should be fixed now thanks to #30819.

I've also adjusted the conversion logic to remove every 2nd note and adjust every 3rd note as that's a much safer approach, because moving the 2nd note would risk it interfering with existing patterns in unintended ways, but moving back the 3rd note would just make it closer to the start of the pattern, therefore making offset changes entirely within the interval of the pattern itself.

> It also turns it into a 0.5 Star Rating

Also got fixed with the improved conversion logic, now that maps convert in a relatively much safer way.

**Comment by Hiviexd** [comment]
Hi, I'm aware that non-critical community PRs don't get much dev time nowadays but I was wondering if this could be granted an exception, as this already got a dev review alongside all issues being fixed. 🙏 

I really do think this mod considerably improves the accessibility of many taiko maps and could be a good reason for certain players to play on lazer, with a bunch of players agreeing with this following feedback back when the PR was made.

**Comment by peppy** [comment]
I"m not against getting it in, but the name has to change.

**Comment by Hiviexd** [comment]
some options are:

- Simplified Rhythm (original name, may be too wordy)
- Simplify (was suggested to me by multiple taiko people)
- Quantize
- Resnap

thoughts? personally leaning towards Simplify.

**Comment by peppy** [comment]
Quantize doesn't work because it's a different concept (and part of the reason I don't like the made-up "quarterize" term – it sounds similar). For the same reason, resnap also doesn't really work.

Simplified Rhythm is pretty nice IMO.

**Comment by Hiviexd** [comment]
Simplified Rhythm it is.

**Comment by peppy** [comment]
@Hiviexd I've done a pass to vastly improve code readability, can you give it a thorough test to make sure I didn't break anything?

**Comment by Hiviexd** [comment]
Everything is good except for 1 issue: the very last pattern of the map doesn't get converted when it needs to.

After thorough testing, I've solved that issue by explicitly handling such case + extracting the conversion logic to its own method.

I've attached the diff below, will commit it if you have no qualms with it.

<details>
<summary>diff</summary>

```diff
diff --git a/osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs b/osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs
index 661e932300..c831fa646d 100644
--- a/osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs
+++ b/osu.Game.Rulesets.Taiko/Mods/TaikoModSimplifiedRhythm.cs
@@ -58,52 +58,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
                     if (inPattern)
                     {
                         // pattern continues
-                        if (snapValue == baseRhythm) continue;
-
-                        inPattern = false;
-
-                        // Iterate through the pattern
-                        for (int j = patternStartIndex; j < i; j++)
+                        if (snapValue == baseRhythm)
                         {
-                            int indexInPattern = j - patternStartIndex;
-
-                            switch (baseRhythm)
+                            // If this is the last note and we're in a pattern, process it
+                            if (i == hits.Length - 1)
                             {
-                                // 1/8: Remove every second note
-                                case 8:
-                                {
-                                    if (indexInPattern % 2 == 1)
-                                    {
-                                        taikoBeatmap.HitObjects.Remove(hits[j]);
-                                    }
-
-                                    break;
-                                }
-
-                                // 1/6 and 1/3: Remove every second note and adjust time of every third
-                                case 6:
-                                case 3:
-                                {
-                                    if (indexInPattern % 3 == 1)
-                                        taikoBeatmap.HitObjects.Remove(hits[j]);
-                                    else if (indexInPattern % 3 == 2)
-                                        hits[j].StartTime = hits[j + 1].StartTime - controlPointInfo.TimingPointAt(hits[j].StartTime).BeatLength / adjustedRhythm;
-
-                                    break;
-                                }
-
-                                default:
-                                    throw new ArgumentOutOfRangeException(nameof(baseRhythm));
+                                inPattern = false;
+                                processPattern(
+                                    hits,
+                                    patternStartIndex,
+                                    i,
+                                    baseRhythm,
+                                    adjustedRhythm,
+                                    controlPointInfo,
+                                    taikoBeatmap
+                                );
                             }
+
+                            continue;
                         }
+
+                        inPattern = false;
+                        processPattern(
+                            hits,
+                            patternStartIndex,
+                            i - 1,
+                            baseRhythm,
+                            adjustedRhythm,
+                            controlPointInfo,
+                            taikoBeatmap
+                        );
                     }
-                    else
+                    else if (snapValue == baseRhythm)
                     {
-                        if (snapValue == baseRhythm)
+                        // For 1/8 patterns, we only need 2 notes
+                        if (baseRhythm == 8)
                         {
                             patternStartIndex = i - 1;
                             inPattern = true;
                         }
+                        // For 1/6 and 1/3, we need at least 3 notes
+                        else if (i >= 2 && getSnapBetweenNotes(controlPointInfo, hits[i - 2], hits[i - 1]) == baseRhythm)
+                        {
+                            patternStartIndex = i - 2;
+                            inPattern = true;
+                        }
                     }
                 }
             }
@@ -114,5 +113,52 @@ private int getSnapBetweenNotes(ControlPointInfo controlPointInfo, Hit currentNo
             var currentTimingPoint = controlPointInfo.TimingPointAt(currentNote.StartTime);
             return controlPointInfo.GetClosestBeatDivisor(currentTimingPoint.Time + (nextNote.StartTime - currentNote.StartTime));
         }
+
+        private void processPattern(
+            Hit[] hits,
+            int startIndex,
+            int endIndex,
+            int baseRhythm,
+            int adjustedRhythm,
+            ControlPointInfo controlPointInfo,
+            TaikoBeatmap taikoBeatmap
+        )
+        {
+            // For 1/8 patterns, we only need ot handle 2 notes
+            // For 1/6 and 1/3 patterns, we need at least 3 notes
+            if (endIndex - startIndex + 1 < (baseRhythm == 8 ? 2 : 3))
+                return;
+
+            for (int j = startIndex; j <= endIndex; j++)
+            {
+                int indexInPattern = j - startIndex;
+
+                switch (baseRhythm)
+                {
+                    case 8:
+                        if (indexInPattern % 2 == 1)
+                            taikoBeatmap.HitObjects.Remove(hits[j]);
+                        break;
+
+                    case 6:
+                    case 3:
+                        // Remove every second note
+                        if (indexInPattern % 3 == 1)
+                        {
+                            taikoBeatmap.HitObjects.Remove(hits[j]);
+                        }
+                        // Move every third note
+                        else if (indexInPattern % 3 == 2 && j < hits.Length - 1)
+                        {
+                            hits[j].StartTime = hits[j + 1].StartTime - controlPointInfo.TimingPointAt(hits[j].StartTime).BeatLength / adjustedRhythm;
+                        }
+
+                        break;
+
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(baseRhythm));
+                }
+            }
+        }
     }
 }
```
</details>

**Comment by peppy** [comment]
I'm trying to fix your odd code quality, so I won't use your diff but thanks for providing it. Do you happen to have a test covering the fail case? That would definitely be useful 😉

**Comment by Hiviexd** [comment]
use this, issue can be spotted in the "1/6 diff" and "1/8 diff" when using their targeted conversions respectively

https://hivie.s-ul.eu/e9koTDEk.osz

**Comment by peppy** [comment]
@Hiviexd please give it another test now 🙏

**Comment by Hiviexd** [comment]
looks solid after further testing, no issues. thanks for fixing this up 👌

**Review by peppy — APPROVED** [review]
